psi4/src/psi4/libfunctional/LibXCfunctional.cc

```diff
@@ -262,106 +254,80 @@ std::map<std::string, double> LibXCFunctional::query_libxc(const std::string& fu
-                                      0.0);
-            failed = false;
-        }
-    } else if (xc_func_name_ == "XC_MGGA_C_TPSS") {
-        if (vsize == 6) {
-            // (xc_func_type *p, double beta, double d, double C0_0, double C0_1, double C0_2, double C0_3);
-            xc_mgga_c_tpss_set_params(xc_functional_.get(), values[0], values[1], values[2], values[3], values[4],
-                                      values[5]);
-            failed = false;
-        }
-    } else if (xc_func_name_ == "XC_MGGA_C_BC95") {
-        if (vsize == 2) {
-            // (XC(func_type) *p, FLOAT css, FLOAT copp);
-            xc_mgga_c_bc95_set_params(xc_functional_.get(), values[0], values[1]);
-            failed = false;
-        }
-        // } else if (xc_func_name_ == "XC_MGGA_C_PKZB") {
-        //     if (vsize == 6) {
-        //         // ((XC(func_type) *p, FLOAT beta, FLOAT d, FLOAT C0_0, FLOAT C0_1, FLOAT C0_2, FLOAT
-        //         // C0_3);
-        //         xc_mgga_c_pkzb_set_params(xc_functional_.get(), values[0], values[1], values[2], values[3],
-        //                                   values[4], values[5]);
-        //         failed = false;
-        //     }
-    } else {
+    int npars = xc_func_info_get_n_ext_params(xc_functional_.get()->info);
+    if (npars == 0) {
         throw PSIEXCEPTION(
             "LibXCfunctional: set_tweak: There are no known tweaks for this functional, please double check "
             "the functional form and add them if required.");
+    } else if (npars != vsize) {
+        std::ostringstream oss;
+        oss << "got " << vsize << ", expected " << npars;
+        throw PSIEXCEPTION(
+            "LibXCfunctional: set_tweak: Mismatch in size of tweaker vector and expected number of "
+            "input parameters:" +
+            oss.str() + "\n");
+    }
+
+    std::vector<double> tweakers_list = values;
+    std::map<std::string, double> tweakers_dict;
+    std::string allowed_keys_join;
+
+    for (int par = 0; par < npars; par++) {
+        std::string key = xc_func_info_get_ext_params_name(const_cast<xc_func_info_type*>(xc_functional_->info), par);
+        allowed_keys_join += key;
+        if (par+1 != npars) allowed_keys_join += ", ";
+        double default_value = xc_func_info_get_ext_params_default_value(const_cast<xc_func_info_type*>(xc_functional_->info), par);
+        if (!quiet) outfile->Printf("Setting parameter #%d (%d/%d) %16s to %16.8f (Libxc default %16.8f).\n", par, par+1, npars, key.c_str(), tweakers_list[par], default_value);
+        tweakers_dict.insert(std::pair<std::string, double>(key, tweakers_list[par]));
     }
 
-    // Did we match fully?
-    if (failed) {
+    outfile->Printf(
+        "Using `LibXCFunctional.set_tweak(std::vector<double>)` instead of `LibXCFunctional.set_tweak(std::map<std::string, double>)` is deprecated, and in 1.5 it will stop working. Allowed keys are: %s\n", allowed_keys_join.c_str());
```

Comment:
Excellent. Maybe chop the long print line onto multiple lines (or if clang-formatter does it, perhaps run it again).